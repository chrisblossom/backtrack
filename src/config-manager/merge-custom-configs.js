/* @flow */

import { getType } from '../utils/get-type';
import { mergeDeep } from '../utils/object-utils';

function mergeCustomConfigs(
    namespace: string,
    config: any,
    customConfigs: $ReadOnlyArray<any>,
) {
    const expectedType = getType(config);
    /**
     * Use reduce right because preset's array order
     */
    const result = customConfigs.reduceRight((acc, currentConfig) => {
        /**
         * Allow custom merge functions to be more explicit how
         */
        if (typeof currentConfig === 'function') {
            let fnResult;

            try {
                fnResult = currentConfig(acc);
            } catch (error) {
                error.message +=
                    // eslint-disable-next-line no-useless-concat
                    '\n' + `config namespace context: ${namespace}`;

                throw error;
            }
            const currentConfigType = getType(fnResult);

            if (currentConfigType !== expectedType) {
                throw new Error(
                    `${namespace} config failed with custom function. Mismatched type. Expected: '${expectedType}', Actual: '${currentConfigType}'`,
                );
            }

            return fnResult;
        }

        const currentConfigType = getType(currentConfig);

        if (currentConfigType !== expectedType) {
            throw new Error(
                `${namespace} config failed. Mismatched type. Expected: '${expectedType}', Actual: '${currentConfigType}'`,
            );
        }

        if (typeof currentConfig === 'object' && currentConfig !== null) {
            if (Array.isArray(currentConfig)) {
                return [...acc, ...currentConfig];
            }

            return mergeDeep(acc, currentConfig);
        }

        return currentConfig;
    }, config);

    return result;
}

export { mergeCustomConfigs };
