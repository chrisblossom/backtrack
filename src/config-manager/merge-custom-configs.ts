/* eslint-disable @typescript-eslint/no-unsafe-return */

import { getType } from '../utils/get-type';
import { mergeDeep } from '../utils/object-utils';

function mergeCustomConfigs(
	namespace: string,
	config: any,
	customConfigs: readonly any[],
) {
	const expectedType = getType(config);
	/**
	 * Use reduce right because preset's array order
	 */
	const result = customConfigs.reduceRight((acc, currentConfig) => {
		/**
		 * Allow custom merge functions to be more explicit how
		 */
		if (typeof currentConfig === 'function') {
			let fnResult;

			try {
				// eslint-disable-next-line @typescript-eslint/no-unsafe-call
				fnResult = currentConfig(acc);
			} catch (error) {
				// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
				error.message +=
					// eslint-disable-next-line no-useless-concat
					'\n' + `config namespace context: ${namespace}`;

				throw error;
			}
			const currentConfigType = getType(fnResult);

			if (currentConfigType !== expectedType) {
				throw new Error(
					`${namespace} config failed with custom function. Mismatched type. Expected: '${expectedType}', Actual: '${currentConfigType}'`,
				);
			}

			return fnResult;
		}

		const currentConfigType = getType(currentConfig);

		if (currentConfigType !== expectedType) {
			throw new Error(
				`${namespace} config failed. Mismatched type. Expected: '${expectedType}', Actual: '${currentConfigType}'`,
			);
		}

		if (typeof currentConfig === 'object' && currentConfig !== null) {
			if (Array.isArray(currentConfig)) {
				return [
					...acc,
					...currentConfig,
				];
			}

			return mergeDeep(acc, currentConfig);
		}

		return currentConfig;
	}, config);

	return result;
}

export { mergeCustomConfigs };
