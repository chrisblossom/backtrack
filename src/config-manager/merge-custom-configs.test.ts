const mergeCustomConfigs = (namespace: any, config: any, customConfigs: any) =>
	require('./merge-custom-configs').mergeCustomConfigs(
		namespace,
		config,
		customConfigs,
	);

const namespace = 'merge_custom_configs_test';

describe('mergeCustomConfigs', () => {
	test('merges objects', () => {
		const config = {
			extends: ['airbnb-base'],
		};

		const customConfigs = [
			{
				extends: ['other-custom-preset'],
			},
			{
				rules: {
					semi: 'always',
				},
			},
			{
				rules: {
					semi: 'never',
				},
			},
		];

		const result = mergeCustomConfigs(namespace, config, customConfigs);

		expect(result).toEqual({
			extends: [
				'airbnb-base',
				'other-custom-preset',
			],
			rules: {
				semi: 'always',
			},
		});
	});

	test('handles custom merge function', () => {
		const config = {
			extends: ['airbnb-base'],
			rules: {
				semi: 'never',
				'no-undef': 'error',
			},
		};

		const customConfigs = [
			(cfg: any) => {
				return {
					...cfg,
					rules: {
						...cfg.rules,
						semi: 'always',
					},
				};
			},
		];

		const result = mergeCustomConfigs(namespace, config, customConfigs);

		expect(result).toEqual({
			extends: ['airbnb-base'],
			rules: { semi: 'always', 'no-undef': 'error' },
		});
	});

	test('handles mixed object and custom merge function', () => {
		const config = {
			extends: ['airbnb-base'],
			rules: {
				semi: 'never',
				'no-undef': 'error',
			},
		};

		const customConfigs = [
			(cfg: any) => {
				return {
					...cfg,
					rules: {
						...cfg.rules,
						semi: 'always',
					},
				};
			},
			{
				rules: {
					semi: 'never',
					'no-undef': 'off',
				},
			},
		];

		const result = mergeCustomConfigs(namespace, config, customConfigs);

		expect(result).toEqual({
			extends: ['airbnb-base'],
			rules: { semi: 'always', 'no-undef': 'off' },
		});
	});

	test('merges arrays', () => {
		const config = ['one'];
		const customConfigs = [
			[
				'two',
				'three',
			],
		];

		const result = mergeCustomConfigs(namespace, config, customConfigs);

		expect(result).toEqual([
			'one',
			'two',
			'three',
		]);
	});

	test('handles string', () => {
		const config = '1';
		const customConfigs = ['2'];

		const result = mergeCustomConfigs(namespace, config, customConfigs);

		expect(result).toEqual('2');
	});

	test('handles numbers', () => {
		const config = 1;

		const customConfigs = [
			2,
			3,
		];

		const result = mergeCustomConfigs(namespace, config, customConfigs);

		expect(result).toEqual(2);
	});

	test('finds invalid preset', () => {
		const config = {};
		const customConfigs = [
			{},
			[],
		];

		let error;
		try {
			mergeCustomConfigs(namespace, config, customConfigs);
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('finds invalid preset function', () => {
		const config = {};
		const customConfigs = [() => ''];

		let error;
		try {
			mergeCustomConfigs(namespace, config, customConfigs);
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('is valid with spread', () => {
		const config = {};

		const customConfigs = [
			(cfg: any) => ({
				...cfg,
				rules: {
					...cfg.rules,
				},
			}),
		];

		const result = mergeCustomConfigs(namespace, config, customConfigs);

		expect(result).toEqual({ rules: {} });
	});

	test('catches thrown errors', () => {
		const config = {};
		const customConfigs = [
			() => {
				throw new Error('broken fn');
			},
		];

		let error;
		try {
			mergeCustomConfigs(namespace, config, customConfigs);
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});
});

// Use an empty export to please Babel's single file emit.
// https://github.com/Microsoft/TypeScript/issues/15230
export {};
