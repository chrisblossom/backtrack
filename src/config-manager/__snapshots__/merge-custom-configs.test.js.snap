// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeCustomConfigs catches thrown errors 1`] = `
[Error: broken fn
config namespace context: merge_custom_configs_test]
`;

exports[`mergeCustomConfigs finds invalid preset 1`] = `[Error: merge_custom_configs_test config failed. Mismatched type. Expected: 'plain object', Actual: 'array']`;

exports[`mergeCustomConfigs finds invalid preset function 1`] = `[Error: merge_custom_configs_test config failed with custom function. Mismatched type. Expected: 'plain object', Actual: 'string']`;
